joint_crab/systematics.py

```diff
@@ -1,333 +1,98 @@
-
-    # sum of all the instrument likelihoods with log-parabola w/ energy scale correction
-    all_instrument_wstat = (
-        wstat_energy_scale(
-            amplitude,
-            reference,
-            alpha,
-            beta,
-            z_fermi,
-            FITTER["spec_obs_lists"]["fermi"],
-            FITTER["bins_in_fit_range"]["fermi"],
-        )
-        + wstat_energy_scale(
-            amplitude,
-            reference,
-            alpha,
-            beta,
-            z_magic,
-            FITTER["spec_obs_lists"]["magic"],
-            FITTER["bins_in_fit_range"]["magic"],
-        )
-        + wstat_energy_scale(
-            amplitude,
-            reference,
-            alpha,
-            beta,
-            z_hess,
-            FITTER["spec_obs_lists"]["hess"],
-            FITTER["bins_in_fit_range"]["hess"],
-        )
-        + wstat_energy_scale(
-            amplitude,
-            reference,
-            alpha,
-            beta,
-            z_veritas,
-            FITTER["spec_obs_lists"]["veritas"],
-            FITTER["bins_in_fit_range"]["veritas"],
-        )
-        + wstat_energy_scale(
-            amplitude,
-            reference,
-            alpha,
-            beta,
-            z_fact,
-            FITTER["spec_obs_lists"]["fact"],
-            FITTER["bins_in_fit_range"]["fact"],
-        )
-        + (z_fermi / delta_fermi) ** 2
-        + (z_magic / delta_magic) ** 2
-        + (z_hess / delta_hess) ** 2
-        + (z_veritas / delta_veritas) ** 2
-        + (z_fact / delta_fact) ** 2
-    )
-    return all_instrument_wstat
-
 
 def main():
     """Fit data with likelihood that includes energy scale uncertainty."""
     log.info("Joint fit including systematics")
-    minuit = run_analysis()
-    compute_fit_results(minuit)
-    compute_contours(minuit)
 
+    log.info("Running fit")
+    fit = SystematicsSpectrumFit()
+    result = fit.run()
+
+    fit_results = make_results_dict(result)
+    utils.write_yaml(fit_results, f"results/fit/fit_joint_systematics.yaml")
+
+    contour_results = compute_contours(fit)
+    utils.write_yaml(contour_results, f"results/fit/contours_joint_systematics.yaml")
+
+
+class SystematicsSpetrumModel(SpectralModel):
+    def __init__(self):
+        self.parameters = Parameters(
+            [
+                Parameter("amplitude", 3e-12, unit="cm-2 s-1 TeV-1"),
+                Parameter("reference", 1, unit="TeV", frozen=True),
+                Parameter("alpha", 2.4, min=1, max=5),
+                Parameter("beta", 0.2, min=0.001, max=1),
+                Parameter("z_fermi", 0),
+                Parameter("z_magic", 0),
+                Parameter("z_veritas", 0),
+                Parameter("z_fact", 0),
+                Parameter("z_hess", 0),
+            ]
+        )
+
+    def make_model(self):
+        model = Log10ParabolaEnergyScale()
+        z = Parameter("z", 0, min=-0.3, max=0.3)
+        parameters = self.parameters.parameters[:4] + [z]
```

Comment:
Can you comment on this? Not sure I understand what it does.

Reply:
There is a global model, and then there is one model object per instrument.
This creates the parameter list of the per-instrument model, which has the first four parameters of the global model linked, and an independent z parameter.

Actually, since I do the update of parameters manually below, I think the whole `make_model` isn't needed, simply putting `return Log10ParabolaEnergyScale()` here would give the same result. Should I try this now, or do you want to take over the systematics analysis now and finish it's scripting / documentation?

Reply:
No, to me it's ok as is.